### DISCUSSION TITLE
Add transparent logo to an existing video

### DISCUSSION THREAD

**Discussion by smixp** [savonet/liquidsoap#2380] [User Support]
Hello, 
I would like to add a transparent logo to an existing video not to a static image. 
this code adds the logo well but I lose the sound of the video. 

```ruby
def logo_overlay(s, logo) =
def mkfilter(graph) =
    background  = ffmpeg.filter.video.input(graph, s)
    logo = ffmpeg.filter.video.input(graph, logo)
    v = ffmpeg.filter.overlay(graph, background, logo, x="1150", y="30")
    ffmpeg.filter.video.output(graph, v)
end
ffmpeg.filter.create(mkfilter)
end
video = logo_overlay(s, logo)
s = mux_video(video=video, blank())
```

Here is my current code
```ruby
url = url_stream
logo = single(url_logo)
s = playlist(url_playlist, mode="normal",reload=1, reload_mode="rounds")

def Add_Logo(~id=null(),logo, s) =
  id =  string.id.default(default="Add_Logo", id)
  s = ffmpeg.raw.encode.audio_video(%ffmpeg(%audio.raw, %video.raw), s)
  v = drop_audio(s)  
  def mkfilter(graph)
    v = ffmpeg.filter.overlay(graph, 
	                          v, 
				  logo, 
				  x="1150", 
				  y="30"
				  )
    ffmpeg.filter.video.output(graph, v)
  end
  
  v = ffmpeg.filter.create(mkfilter)
  s = mux_audio(id=id,audio=drop_video(s),v)
  
  ffmpeg.raw.decode.audio_video(s)
end

s = Add_Logo(logo, s)

encoder = %ffmpeg(format="flv",
                  %audio(codec="aac", b="128k", samplerate=48000, q=4),
		  %video.raw(codec="libx264", width=1280, height=720, b="3000k", r="29", g="48", "x264-params"="scenecut=0:open_gop=0:min-keyint=150:keyint=150", preset="veryfast")
				  )
				  
output.url(fallible=true, url=url, encoder, s)
```

Erreur line 13 => v
```/liquidsoap\_Local_ffmpeg.liq, line 13, char 27:
                                 v,
Error 5: this value has type
source(_) (inferred at /liquidsoap/_Local_ffmpeg.liq, line 9, char 6-19)
but it should be a subtype of
  ffmpeg.filter.video```

thank you in advance for your help

**Comment by toots** [comment]
Hi!

Couple of things:
* You don't need to do the inline encoding, the ffmpeg decoder is capable of decoding directly into the ffmpeg raw format.
* You need to convert sources to ffmpeg filter inputs and ffmpeg filter outputs back to sources when using filters.
* The ffmpeg encoder should also be able to deal with raw audio and video
* There is currently a limitation on clocks so you need to also pass the audio part through the filter.

Adapted script, seemed to be working here with the latest `main`:
```ruby
url = url_stream
logo = single(url_logo)
s = playlist(url_playlist, mode="normal",reload=1, reload_mode="rounds")

def Add_Logo(~id=null(),logo,s) =
  id =  string.id.default(default="Add_Logo", id)
  def mkfilter(graph)
    background  = ffmpeg.filter.video.input(graph, drop_audio(s))
    logo = ffmpeg.filter.video.input(graph, logo)
    v = ffmpeg.filter.overlay(graph,
                                  background,
                                  logo,
                                  x="1150",
                                  y="30"
                                  )
    v = ffmpeg.filter.video.output(graph, v)

    # Remove when https://github.com/savonet/liquidsoap/issues/2384 is closed
    a = ffmpeg.filter.audio.input(graph, drop_video(s))
    a = ffmpeg.filter.anull(graph, a)
    a = ffmpeg.filter.audio.output(graph, a)

    (a, v)
  end

  let (a, v) = ffmpeg.filter.create(mkfilter)
  mux_audio(id=id,audio=a,v)
end

s = Add_Logo(logo, s)

encoder = %ffmpeg(format="flv",
                  %audio.raw(codec="aac", b="128k", samplerate=48000, q=4),
		  %video.raw(codec="libx264", width=1280, height=720, b="3000k", r="29", g="48", "x264-params"="scenecut=0:open_gop=0:min-keyint=150:keyint=150", preset="veryfast")
				  )
				  
output.url(fallible=true, url=url, encoder, s)
```

**Reply by smixp** [reply]
Hello, 
Thank you again for your help 
Your code works, but in less than 5 minutes it occupies all the memory (16GB) and it shuts down. 
In the logs, it repeats this line until the crash.

Code:
```ruby
url = url_stream
logo = single(url_logo)
s = playlist(url_playlist, mode="normal",reload=1, reload_mode="rounds")

def Add_Logo(~id=null(),logo,s) =
  id =  string.id.default(default="Add_Logo", id)
  def mkfilter(graph)
    background  = ffmpeg.filter.video.input(graph, drop_audio(s))
    logo = ffmpeg.filter.video.input(graph, logo)
    v = ffmpeg.filter.overlay(graph,
                                  background,
                                  logo,
                                  x="1150",
                                  y="30"
                                  )
    v = ffmpeg.filter.video.output(graph, v)
    a = ffmpeg.filter.audio.input(graph, drop_video(s))
    a = ffmpeg.filter.anull(graph, a)
    a = ffmpeg.filter.audio.output(graph, a)
    (a, v)
  end
  let (a, v) = ffmpeg.filter.create(mkfilter)
  mux_audio(id=id,audio=a,v)
end

s = Add_Logo(logo, s)

encoder = %ffmpeg(format="flv",
                  %audio.raw(codec="aac", b="128k", samplerate=48000, q=4),
		  %video.raw(codec="libx264", width=1280, height=720, b="3000k", r="29", g="48", "x264-params"="scenecut=0:open_gop=0:min-keyint=150:keyint=150", preset="veryfast")
				  )
output.url(fallible=true, url=url, encoder, s)
```

Logs :
```
2022/05/08 02:24:03 [_playlist.txt:4] Queued 1 requests
[libx264 @ 0000020024ab7500] specified frame type (3) at 0 is not compatible with keyframe interval
[libx264 @ 0000020024ab7500] forced frame type (5) at 11 was changed to frame type (3)
[libx264 @ 0000020024ab7500] forced frame type (5) at 31 was changed to frame type (3
[libx264 @ 0000020024ab7500] forced frame type (5) at 52 was changed to frame type (3)
[abuffersink_0 @ 0000020025277300] 100 buffers queued in abuffersink_0, something may be wrong.
[libx264 @ 0000020024ab7500] forced frame type (5) at 72 was changed to frame type (3)
[libx264 @ 0000020024ab7500] forced frame type (5) at 78 was changed to frame type (3)
[libx264 @ 0000020024ab7500] forced frame type (5) at 83 was changed to frame type (3)
```

I did some research and apparently it comes from "**pict_type**". 
Here is the solution I found but I don't know how to adapt it to my code
Solution:

_Since I was just trying to transcode the data, I was simply taking pictures that had been emitted by the decoder and passing them to the encoder. However, the decoder sets the pict_type value for the AVFrame structures it produces to AV_PICTURE_TYPE_P. The encoder, in turn, uses the pict_type value to determine the type of frames to generate. Hence, the encoder only ever produces P-frames since that is what is always specified in the input.
Resetting the **pict_type** field to **AV_PICTURE_TYPE_NONE** allows the encoder to produce whatever type of frame it thinks is best. Once I updated my code to do this, my test utility generated B-frames and produced a file identical to the one output by ffmpeg itself._

thank you in advance for your help

**Reply by toots** [reply]
I'm not sure of this solution is related to your problem. I would suggest to start by removing all encoder params and see if that works better:
```ruby
encoder = %ffmpeg(
  format="flv",
  %audio.raw(codec="aac"),
  %video.raw(codec="libx264")
)

output.url(fallible=true, url=url, encoder, s)
```
Then, if that works, you can start introducing them one by one.

I would also double check which version of `ocaml-ffmpeg` you are using. Version `1.1.3`, the last release currently, fixed some important issues with how OCaml handles the memory allocated by `ffmpeg`.

**Reply by smixp** [reply]
Hello,  
I deleted all encoder settings and it still works the same (in less than 7 minutes the memory is full and crash:

I am on windows 10, 
Processor: Intel(R) Core(TM) i7-6500U CPU @ 2.50GHz   2.60 GHz
Ram:  16.0 GB (15.9 GB usable)
Liquidsoap version : 2.0.4
Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.4.6 sedlex=2.2 menhirLib=20210929 dtools=0.4.4 duppy=0.9.2 cry=0.6.6 mm=0.8.0 xmlplaylist=0.1.5 dynlink=[distributed with Ocaml] ffmpeg=1.1.2 ao=0.2.1 samplerate=0.1.6 taglib=0.3.7 ssl=0.5.9 camomile=1.0.2 yojson=1.7.0 portaudio=0.2.3 srt.constants=0.2.1 srt.types=0.2.1 srt.stubs=0.2.1 srt.stubs.locked=0.2.1 srt=0.2.1 winsvc=1.0.1
The processor and the ram are both at 100% (16GO)  in less than 7 minutes. I don't know where the problem comes from 

Logs :
`
[libx264 @ 000002c71b198ec0] forced frame type (5) at 5930 was changed to frame type (3)
2022/05/08 04:03:54 [clock.main:2] We must catchup 1.02 seconds!
2022/05/08 04:03:55 [clock.main:2] We must catchup 1.03 seconds!
2022/05/08 04:03:56 [clock.main:2] We must catchup 1.39 seconds!
2022/05/08 04:03:57 [clock.main:2] Too much latency! Resetting active sources...
2022/05/08 04:04:00 [clock.main:2] We must catchup 1.07 seconds!
2022/05/08 04:04:02 [clock.main:2] Too much latency! Resetting active sources...
2022/05/08 04:04:08 [clock.main:2] We must catchup 1.92 seconds!
2022/05/08 04:04:09 [clock.main:2] Too much latency! Resetting active sources...
2022/05/08 04:04:10 [clock.main:2] We must catchup 1.27 seconds!
2022/05/08 04:04:11 [clock.main:2] We must catchup 1.44 seconds!
2022/05/08 04:04:13 [clock.main:2] We must catchup 1.66 seconds!
2022/05/08 04:04:13 [clock.main:2] Too much latency! Resetting active sources...
[libx264 @ 000002c71b198ec0] forced frame type (5) at 6330 was changed to frame type (3)
2022/05/08 04:04:17 [clock.main:2] We must catchup 1.03 seconds!
2022/05/08 04:04:19 [clock.main:2] We must catchup 1.58 seconds!
2022/05/08 04:04:20 [clock.main:2] Too much latency! Resetting active sources...
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Source video_main_0 failed while streaming: Avutil.Error(Cannot allocate memory)!
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Raised by primitive operation at Avfilter.launch.(fun) in file "avfilter/avfilter.ml", line 392, characters 33-61
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Ffmpeg_filter_io.video_input#flush_buffer.(fun) in file "io/ffmpeg_filter_io.ml", line 289, characters 27-53
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Ffmpeg_filter_io.input_base#pull.(fun).f in file "io/ffmpeg_filter_io.ml", line 149, characters 19-27
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Stdlib__option.map in file "option.ml", line 24, characters 57-62
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Ffmpeg_filter_io.input_base#pull in file "io/ffmpeg_filter_io.ml", line 143, characters 8-462
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Ffmpeg_filter_io.input_base#get_frame in file "io/ffmpeg_filter_io.ml", line 164,
characters 65-74
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Source.operator#instrumented_get_frame in file "source.ml", line 668, characters 6-24
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Source.operator#get in file "source.ml", line 739, characters 10-4
2022/05/08 04:04:25 [clock.mux_audio_0.child:4] Called from Output.output#output in file "outputs/output.ml", line 165, characters 10-30`

I tried with the external encoder ffmpeg and it works very well with the same parameters: processor result: 25% RAM 25% (270 MB maximum)
`cmd = "#{ffmpeg}  -y -thread_queue_size 2024  -i pipe:0 -i #{logo} -filter_complex \"overlay=x=main_w-overlay_w-(main_w*0.05):y=main_h*0.05\" -vcodec libx264 -preset veryfast -crf 29 -r 30 -g 48 -b:v 3000k -s 1280x720 -acodec libmp3lame -ar 44100-b:a 128k -tune zerolatency -f flv -flvflags no_duration_filesize #{url}"
output.external(%avi(), cmd, s)`

I just tested with Liquidsoap 2.0.5+dev and it's still the same memory issue.
I tried without the Add_Logo(logo,s) function and the normal 300MB memory
I guess the problem comes from the ffmpeg filters because without this function (add_logo) everything works fine.
there must be a repeating loop somewhere

thank you in advance for your help
